### PR TITLE
refactor(console): improve stepper service

### PR DESCRIPTION
**Pull request by ThibaudAV**
## Issue

https://gravitee.atlassian.net/browse/APIM-454

## Description

- Add dedicated service for step ApiCreationStepService
- Change payload to have patch function

## Additional context

🚀 CI was able to deploy the build of this PR, so you can now try it directly [here](https://apimnightlywebui24386.z6.web.core.windows.net/apim-569-stepper/index.html)
_Notes_: The deployed app is linked to the management API of APIM master. (Same login and password as APIM master)

---

📚&nbsp;&nbsp;View the storybook of this branch [here](https://612657caa8e859003a8a6430-krltlrcgjv.chromatic.com)

### CODE REVIEW

File: gravitee-apim-console-webui/src/management/api/creation/v4/models/ApiCreationStepService.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import { Injectable } from '@angular/core';
+
+import { ApiCreationPayload } from './ApiCreationPayload';
+import { ApiCreationStep, ApiCreationStepperService } from './ApiCreationStepperService';
+
+/**
+ * This service is injected in each step component to provide a way to access the payload and to navigate between steps.
+ */
+@Injectable()
+export class ApiCreationStepService {
```
Comment (loriepisicchio) [inline gravitee-apim-console-webui/src/management/api/creation/v4/models/ApiCreationStepService.ts]: The file naming does not match usual conventions ApiCreationStepService.ts should be api-creation-step.service.ts.
Same for ApiCreationPayload and ApiCreationStepperService

File: gravitee-apim-console-webui/src/management/api/creation/v4/models/ApiCreationStepperService.ts [hunk 2]
```diff
@@ -13,29 +13,31 @@
...
+  position: number;
   label: string;
   component: Type<unknown>;
-  payload: ApiCreationPayload;
+  patchPayload: (lastPayload: ApiCreationPayload) => ApiCreationPayload;
   type: 'primary' | 'secondary';
 }
 
-type ApiCreationStepPayload = Omit<ApiCreationStep, 'type' | 'payload'> & { payload?: ApiCreationPayload };
+type ApiCreationStepPayload = Omit<ApiCreationStep, 'type' | 'patchPayload'>;
```
Comment (loriepisicchio) [inline gravitee-apim-console-webui/src/management/api/creation/v4/models/ApiCreationStepperService.ts]: The `ApiCreationStepPayload` looks a bit confusing to me, I had a hard time understanding its use. WDYT about `ApiCreationStepDescriptor` instead ? 

And also splitting in 2 types and having an intersection type like follows would make it easier to understand I think
```
export interface ApiCreationStepDescriptor {
  position: number;
  label: string;
  component: Type<unknown>;
}

export interface ApiCreationStepPayload {
  patchPayload: (lastPayload: ApiCreationPayload) => ApiCreationPayload;
  type: 'primary' | 'secondary';
}

export type ApiCreationStep = ApiCreationStepDescriptor & ApiCreationStepPayload;
```

File: gravitee-apim-console-webui/src/management/api/creation/v4/models/ApiCreationStepperService.ts [hunk 3]
```diff
@@ -48,35 +50,45 @@ export class ApiCreationStepperService {
...
+   * @param {ApiCreationStep} step - the current step of the API creation process
+   * @returns {ApiCreationPayload} - the data for the current step
+   */
+  public compileStepPayload(step: ApiCreationStep): ApiCreationPayload {
+    // Get all the previous steps of the current step
+    const previousSteps = this.steps.slice(0, this.steps.indexOf(step));
+    // Compute the data for the current step by using reduce method on the previous steps
+    // The reduce method is used to combine all the data from the previous steps using the patchPayload method
+    // The initial payload is used as the starting value for the reduce method
+    return step.patchPayload(previousSteps.reduce((payload, previousStep) => previousStep.patchPayload(payload), this.initialPayload));
```
Comment (jourdiw) [inline gravitee-apim-console-webui/src/management/api/creation/v4/models/ApiCreationStepperService.ts]: To simplify this line, create a list with previous steps and the 'step' parameter at the end, then reduce.

File: gravitee-apim-console-webui/src/management/api/creation/v4/models/ApiCreationStepService.ts (hunk 1 above)
Comment (ThibaudAV) [reply gravitee-apim-console-webui/src/management/api/creation/v4/models/ApiCreationStepService.ts]: yes done,  now is not juste a class but a service

File: gravitee-apim-console-webui/src/management/api/creation/v4/models/ApiCreationStepperService.ts (hunk 2 above)
Comment (ThibaudAV) [reply gravitee-apim-console-webui/src/management/api/creation/v4/models/ApiCreationStepperService.ts]: Hum,  The idea of the type ApiCreationStepPayload it is really just to init the stepper with a simplified ApiCreationStep. and it's not exported only for constructor use 🤷‍♂️ 
`type` are also not require

**Review by loriepisicchio — APPROVED** [review]

**Pull request merged**